From 1150fdc4652eb677fceede9e5383d789572853be Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Wed, 22 Oct 2014 22:57:17 -0400
Subject: [powerpc] vphn: NUMA node code expects big-endian

Message-id: <20141022225717.9827.44079.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 98089
O-Subject: [PATCH RHEL7.1 BZ1154673] powerpc/vphn: NUMA node code expects big-endian
Bugzilla: 1154673
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1154673

Description:
------------
The associativity domain numbers are obtained from the hypervisor through
registers and written into memory by the guest: the packed array passed to
vphn_unpack_associativity() is then native-endian, unlike what was assumed
in the following commit:

commit b08a2a12e44eaec5024b2b969f4fcb98169d1ca3
Author: Alistair Popple <alistair@popple.id.au>
Date:   Wed Aug 7 02:01:44 2013 +1000

    powerpc: Make NUMA device node code endian safe

This issue fills the topology with bogus data and makes it unusable. It may
lead to severe performance breakdowns.

We should ideally patch the vphn_unpack_associativity() function to do the
64-bit loads, but this requires some more brain storming.

In the meantime, let's go for a suboptimal and temporary bug fix: this patch
converts each 64-bit value of the packed array to big endian, as expected by
the current parsing code in vphn_unpack_associativity().

Signed-off-by: Greg Kurz <gkurz@linux.vnet.ibm.com>
Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>

RHEL Version Found:
-------------------
RHEL 7.0

Brew:
-----
BE: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8148580
LE: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8148591

Upstream:
---------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=5c9fb1899400096c6818181c525897a31d57e488

Test Status:
------------
Tested on both power BE and power LE systems.
Ran Kernel Tier 1 successfully on BE system (ibm-p740-01-lp5): https://beaker.engineering.redhat.com/jobs/781576

---------------------------------------------------------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/mm/numa.c b/arch/powerpc/mm/numa.c
index 141a8a8..2d9f624 100644
--- a/arch/powerpc/mm/numa.c
+++ b/arch/powerpc/mm/numa.c
@@ -1469,8 +1469,11 @@ static long hcall_vphn(unsigned long cpu, __be32 *associativity)
  long retbuf[PLPAR_HCALL9_BUFSIZE] = {0};
  u64 flags = 1;
  int hwcpu = get_hard_smp_processor_id(cpu);
+ int i;
 
  rc = plpar_hcall9(H_HOME_NODE_ASSOCIATIVITY, retbuf, flags, hwcpu);
+ for (i = 0; i < 6; i++)
+  retbuf[i] = cpu_to_be64(retbuf[i]);
  vphn_unpack_associativity(retbuf, associativity);
 
  return rc;
-- 
1.7.1